### PR TITLE
Working skew

### LINKED ISSUE #726: Add support for skewing [$125]
As discussed in #95, this is the issue to support skewing, with the bounty attached

Things that need fixing:
1. Skewing manual by controls / programmatically
2. Skewing due to side / top stretching a group containing a rotated object

<bountysource-plugin>
---
There is a **[$125 open bounty](https://www.bountysource.com/issues/440471-add-support-for-skewing-125?utm_campaign=plugin&utm_content=tracker%2F23217&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F23217&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by asturur**
I put it here so i know i have to complete it.

closes #726

### CODE REVIEW

**Comment by asturur** [comment]
@kangax This is a working skew PR. toSvg() is still missing because i have another PR that is waiting to be merged on the same file.
Could you please review the code?

Here a working demo:
http://www.deltalink.it/andreab/fabric/kitchensink.html
paste this in execute tab:
```
// clear canvas
canvas.clear();

// remove currently selected object
canvas.remove(canvas.getActiveObject());

// add red rectangle
canvas.add(new fabric.Rect({
  width: 50,
  height: 50,
  left: 50,
  top: 50,
  fill: 'rgb(255,0,0)',
  skewX: 30
}));

// add red rectangle
canvas.add(new fabric.Rect({
  width: 50,
  height: 50,
  left: 50,
  top: 50,
  fill: 'rgb(0,250,0)',
  skewX: 30,
  skewY: -30
}));

// add green, half-transparent circle
canvas.add(new fabric.Circle({
  radius: 40,
  left: 50,
  top: 50,
  fill: 'rgb(0,255,0)',
  opacity: 0.5,
  skewY: 60
}));
```

**Comment by asturur** [comment]
it does not build on travis, i do not know why.
It looks like it is failing the minification.

**Comment by asturur** [comment]
Added background work to add mouse operation skewing.
Actual skewing calculation is still missing.
Mirrored properties lockScaleX, losckScaleY, centeredScaling also for skewing.

Assigne skewing action with SHIFT + middle corners.

Alt + middle, alt + (tl,br, bl ,tr), shift + (tl,br, bl ,tr) were already taken.
Ctrl key is missing on mac and i read is hard to use the cmd key good in the browsers.
(correct me if i'm wrong, changing key is very easy ).

File: src/mixins/object_interactivity.mixin.js [hunk 1]
```diff
@@ -142,8 +142,30 @@
       if (!dimensions) {
         dimensions = this._getNonTransformedDimensions();
       }
-      var transformMatrix = this._calcDimensionsTransformMatrix();
-      return fabric.util.transformPoint(dimensions, transformMatrix, true);
+      var points = [
+          {
+            x: -dimensions.x/2,
```
Comment (kangax) [inline src/mixins/object_interactivity.mixin.js]: Maybe cache these?

Comment (asturur) [inline src/mixins/object_interactivity.mixin.js]: maybe is better to cache the dimensions and matrix with the combinatio of skew and scale.
something like we do with font width caching and beizer to arc caching.

but as of now is an early optimization. help me to understand why it fails minifing

**Comment by kangax** [comment]
No idea why minification fails...

**Comment by asturur** [comment]
Added skew by mouse, with some defects.
you can try my linked kitchensink dragging the medium corners while holding shift.

It works mixed with scale, it does not work good if you intentionally pull angles in strange situation.

As of now skewX or skewY is SET to certain value, while we should calculate a skew difference with previous transform and add it to the current skew.

**Comment by asturur** [comment]
Just to clarify all tests from adjustPosition are failing because now borderWidth is included in the calculation.
This does not effect rendering but i do not understand why.
http://www.deltalink.it/andreab/fabric/kitchensink.html

**Comment by asturur** [comment]
Dilemma:
image

take an svg object, skewX and then skewY gives different results than doing the opposite skewY and then skewX.
It means that we have to do two differents transforms for that, and this is not a problem.

But which order should we choose?
X and then Y or viceversa?

**Comment by asturur** [comment]
Added transform.shiftKey , transform.altKey, removed event paramenter from shouldCenterTransform and resetCurrentTransform.

**Comment by asturur** [comment]
I just ruined the branch, and now? how do i unmerge a merge?

**Pull request closed without merging**